cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,387 @@
+		}
+		return tx, false, ledgerRange
+	}
+
+	//
+	// On-the-fly ingestion: extract all of the fields, return best effort.
+	//
+	tx.FeeBump = ingestTx.Envelope.IsFeeBump()
+	tx.ApplicationOrder = int32(ingestTx.Index)
+	tx.Successful = ingestTx.Result.Successful()
+	tx.Ledger = ledgerbucketwindow.LedgerInfo{
+		Sequence:  lcm.LedgerSequence(),
+		CloseTime: lcm.LedgerCloseTime(),
+	}
+
+	if tx.Result, err = ingestTx.Result.Result.MarshalBinary(); err != nil {
+		log.WithError(err).Errorf("Failed to encode transaction Result")
+		return tx, false, ledgerRange
+	}
+	if tx.Meta, err = ingestTx.UnsafeMeta.MarshalBinary(); err != nil {
+		log.WithError(err).Errorf("Failed to encode transaction UnsafeMeta")
+		return tx, false, ledgerRange
+	}
+	if tx.Envelope, err = ingestTx.Envelope.MarshalBinary(); err != nil {
+		log.WithError(err).Errorf("Failed to encode transaction Envelope")
+		return tx, false, ledgerRange
+	}
+	if events, diagErr := ingestTx.GetDiagnosticEvents(); diagErr == nil {
+		tx.Events = make([][]byte, 0, len(events))
+		for i, event := range events {
+			bytes, ierr := event.MarshalBinary()
+			if ierr != nil {
+				log.WithError(ierr).
+					Errorf("Failed to encode transaction DiagnosticEvent %d", i)
+				return tx, false, ledgerRange
+			}
+			tx.Events = append(tx.Events, bytes)
+		}
+	} else {
+		log.WithError(diagErr).
+			Errorf("Failed to encode transaction DiagnosticEvents")
+		return tx, false, ledgerRange
+	}
+
+	log.WithField("txhash", hexHash).
+		WithField("duration", time.Since(start)).
+		Debugf("Fetched and encoded transaction from ledger %d",
+			lcm.LedgerSequence())
+
+	return tx, true, ledgerRange
+}
+
+// getTransactionByHash actually performs the DB ops to cross-reference a
+// transaction hash with a particular set of ledger close meta and parses out
+// the relevant transaction efficiently by leveraging the `application_order` db
+// field.
+//
+// Note: Caller must do input sanitization on the hash.
+func (txn *transactionReaderTx) getTransactionByHash(hash string) (
+	xdr.LedgerCloseMeta, ingest.LedgerTransaction, error,
+) {
+	rowQ := sq.Select("t.application_order", "lcm.meta").
+		From(fmt.Sprintf("%s t", transactionTableName)).
+		Join(fmt.Sprintf("%s lcm ON (t.ledger_sequence = lcm.sequence)", ledgerCloseMetaTableName)).
+		Where(sq.Eq{"t.hash": hash}).
+		Limit(1).
+		RunWith(txn.db.GetTx())
+
+	rows, err := rowQ.Query()
+	if err != nil {
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+			errors.Wrapf(err, "failed row read for txhash %s", hash)
+	}
+	if !rows.Next() {
+		if ierr := rows.Err(); ierr != nil {
+			return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{},
+				errors.Wrapf(ierr, "failed to fetch txhash %s", hash)
+		}
+
+		// We use EOF to mark "this tx doesn't exist," since that isn't a real error.
+		return xdr.LedgerCloseMeta{}, ingest.LedgerTransaction{}, io.EOF
+	}
+
+	var (
+		txIndex int
+		meta    []byte
+	)
+	if err := rows.Scan(&txIndex, &meta); err != nil {
```

Comment:
xdr.LedgerCloseMeta implements a db scan function so you can scan an instance of xdr.LedgerCloseMeta here instead of manually unmarshalling the binary a few lines below

Reply:
nice tip!